### PR TITLE
Prevent `<style>` from getting moved in `<html>` 

### LINKED ISSUE #712: 🐛 BUG: [AST] Style tags get merged into html tag in certain situations
### What version of `@astrojs/compiler` are you using?

1.0.1

### What package manager are you using?

pnpm

### What operating system are you using?

macOS

### Describe the Bug

In some cases, style tags gets merged into `html`. It seems to depends on the content of `head`, normal content works, but having a component there cause an issue. Cause the following issue in the Prettier plugin: https://github.com/withastro/prettier-plugin-astro/issues/316

Minimum repro:
```astro
<html lang="en">
<head>
    <BaseHead />
</head>
</html>

<style>
    @use "../styles/global.scss";
</style>
```

gets formatted into:

```astro
<html lang="en">
  <head>
    <BaseHead />
  </head>
  <style>
    @use "../styles/global.scss";
  </style>
</html>

```

Result AST: https://gist.github.com/Princesseuh/7ac0827647d2ff540a7bcadc7e8c11dc

### Link to Minimal Reproducible Example

See description

### PR DESCRIPTION
**Pull request by MoustaphaDev**
## Changes

- Closes #712 

## Testing

Added json test to ensure a correct AST
Added tsx tests to ensure a correct result

## Docs

N/A bug fix only

### CODE REVIEW

File: internal/parser.go [hunk 1]
```diff
@@ -902,8 +902,10 @@ func inHeadIM(p *parser) bool {
 			p.im = afterHeadIM
 			return true
 		case a.Body, a.Html, a.Br:
-			p.parseImpliedToken(EndTagToken, a.Head, a.Head.String())
```
Comment (MoustaphaDev) [inline internal/parser.go:905]: The fix was to remove this divergence from the [html spec](https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inhead:~:text=tag%20name%20is%20one%20of%3A%20%22body%22%2C%20%22html%22%2C%20%22br%22-,Act%20as%20described%20in%20the%20%22anything%20else%22%20entry%20below.,-A%20start%20tag%20whose%20tag%20name%20is%20%22template) when in the `inHead` insertion mode, see the ["any thing"](https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inhead:~:text=Ignore%20the%20token.-,Anything%20else,-Pop%20the%20current) section to find the correct steps to be performed

**Review by natemoo-re — COMMENTED** [review]
Is this ready for review?

**Comment by MoustaphaDev** [comment]
I wanted to get more tests to cover all issues opened about this bug, but besides that yeah. I can't work on it right now though unfortunately

**Comment by MoustaphaDev** [comment]
I added a fair number of tests though so feel free to push it forward!

**Review by Princesseuh — APPROVED** [review]

**Pull request merged**